executor/aggfuncs/func_lead_lag.go

```diff
@@ -0,0 +1,91 @@
+// Copyright 2019 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package aggfuncs
+
+import (
+	"github.com/pingcap/tidb/expression"
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/types"
+	"github.com/pingcap/tidb/util/chunk"
+)
+
+type baseLeadLag struct {
+	baseAggFunc
+
+	tp          *types.FieldType
+	defaultExpr expression.Expression
+	offset      uint64
+}
+
+type partialResult4LeadLag struct {
+	evaluator valueEvaluator
```

Comment:
evaluator should be put in `baseLeadLag`? The state stored in it is only used in `AppendFinalResult2Chunk`

Reply:
It has some internal status that will be changed, so as a convention, it should not be put in `baseLeadLag` because it is better read-only.

Reply:
The lifecycle of these internal status won't beyond `AppendFinalResult2Chunk()`?

Reply:
Yes.

Reply:
IMO, these status can be regarded as temporary variables used only in `AppendFinalResult2Chunk()`. In the outside of the window function evaluation scope, like WindowExec, this evaluator is stateless. So I think it can be put in the `baseLeadLag`.

Reply:
Currently yes. But t would still be put in the partial result if we want to parallelize the window exec.

Reply:
How about moving it into the partial result structure in the future when we need it. We can add a TODO like:
```go
type baseLeadLag struct {
	baseAggFunc
	valueEvaluator // TODO: move it to partial result when parallel execution is supported

 	tp          *types.FieldType
	defaultExpr expression.Expression
	offset      uint64
}
```